File: priv/scripts/generate_proto_file.py
#!/usr/bin/env nix-shell
#! nix-shell -i python3 -p python3
import json
import sys
from collections import OrderedDict
from pathlib import Path
from pprint import pprint


def override_type(parent_name, field_name, current_type):
    if parent_name != "Wrapper" and current_type == "bytes":
        return "Wrapper"
    elif parent_name == "Error" and field_name == "code":
        return "ErrorCode"
    else:
        return current_type


def format_line(line, indent):
    return f"{' ' * 2 * indent}{line}\n"


def parse_rule(field):
    if "rule" in field:
        return field["rule"] + " "
    else:
        return ""


def parse_fields(message, fields, indent):
    return (
        format_line(
            f"{parse_rule(field)}{override_type(message, name, field['type'])} {name} = {field['id']};",
            indent,
        )
        for name, field in fields.items()
    )


def parse_methods(methods, indent):
    return (
        format_line(
            f"rpc {name} ({method['requestType']}) returns ({method['responseType']});",
            indent,
        )
        for name, method in methods.items()
    )


def parse_values(values, indent):
    return (format_line(f"{name} = {value};", indent) for name, value in values.items())


def parse_item(name, item, indent):
    lines = []

    if "fields" in item:
        lines = [
            format_line(f"message {name} {{", indent),
            *parse_fields(name, item["fields"], indent + 1),
        ]
    elif "methods" in item:
        lines = [
            format_line(f"service {name} {{", indent),
            *parse_methods(item["methods"], indent + 1),
        ]
    elif "values" in item:
        lines = [
            format_line(f"enum {name} {{", indent),
            *parse_values(item["values"], indent + 1),
        ]
    else:
        pprint(item)
        raise Exception("Unrecognized Item")

    if "nested" in item:
        assert len(item) == 2
        for name, item in item["nested"].items():
            lines.extend(parse_item(name, item, indent + 1))

    lines.append(format_line("}\n", indent))

    return lines


def parse(input_file):
    data = json.load(input_file.open("r"), object_pairs_hook=OrderedDict)
    data_nested = data["nested"]
    assert len(data_nested) == 1
    package_name = list(data_nested.keys())[0]

    lines = [
        'syntax = "proto3";\n\n',
        f"package {package_name};\n\n",
        'import "error.proto";\n\n',
    ]

    package = data_nested[package_name]
    package_nested = package["nested"]

    for name, item in package_nested.items():
        lines.extend(parse_item(name, item, 0))

    return lines


def main():
    input_file = Path(sys.argv[1])
    output_file = Path(sys.argv[2])
    lines = parse(input_file)

    with output_file.open("w") as f:
        f.writelines(lines)


if __name__ == "__main__":
    main()

File: priv/scripts/generate_proto_file.exs
defmodule GenerateProtoFile do
  def run([input_file, output_file]) do
    input_file
      |> File.read!()
      |> Jason.decode!(objects: :ordered_objects)
      |> parse()
      |> indent_lines()
      |> Enum.join("\n")
      |> (&File.write!(output_file, &1)).()

    IO.puts "Converted '#{input_file}' to '#{output_file}'"
  end

  def indent_lines(lines) do
    Enum.map(
      lines,
      fn {text, depth} ->
        String.duplicate(" ", 2 * depth) <> text
      end
    )
  end

  def parse(input) do
    depth = 0
    nested = input["nested"]
    {package_name, package} = Enum.at(nested, 0)

    header = [
        {"syntax = \"proto3\";\n", depth},
        {"package #{package_name};\n", depth},
    ]

    lines = Enum.flat_map(
      package["nested"],
      fn {name, item} -> parse_items(name, item, depth) end
    )

    header ++ extra_headers() ++ lines
  end

  def parse_items(name, item, depth) do
    parsed_items =
      Enum.flat_map(
        item,
        fn {kind, values} ->
          parse_item(kind, name, values, depth) end
      )

    end_block = [{"}\n", depth}]

    parsed_items ++ end_block
  end

  def parse_item("fields", name, values, depth) do
    header = {"message #{name} {", depth}
    [header | parse_fields(name, values, depth + 1)]
  end

  def parse_item("methods", name, values, depth) do
    header = {"service #{name} {", depth}
    [header | parse_methods(values, depth + 1)]
  end

  def parse_item("values", name, values, depth) do
    header = {"enum #{name} {", depth}
    [header | parse_values(values, depth + 1)]
  end

  def parse_item("nested", _name, items, depth) do
    Enum.flat_map(
      items,
      fn {name, values} ->
        parse_items(name, values, depth + 1) end
    )
  end

  def parse_item(kind, _name, _values, _depth) do
    raise "Unknown item kind #{kind}"
  end

  def parse_values(values, depth) do
    Enum.map(values, fn {name, value} ->
      {"#{name} = #{value};", depth}
    end)
  end

  def parse_methods(methods, depth) do
    Enum.map(methods, fn {name, method} ->
      request_type = method["requestType"]
      response_type = method["responseType"]
      text = "rpc #{name} (#{request_type}) returns (#{response_type});"
      {text, depth}
    end)
  end

  def parse_fields(message, fields, depth) do
    Enum.map(fields, fn {name, field} ->
      rule = parse_rule(field)
      new_type = maybe_override_type(message, name, field["type"])
      text = "#{rule}#{new_type} #{name} = #{field["id"]};"
      {text, depth}
    end)
  end

  def parse_rule(field) do
    case field["rule"] do
      nil -> ""
      rule -> "#{rule} "
    end
  end

  def maybe_override_type(parent_name, field_name, current_type) do
    cond do
      parent_name != "Wrapper" and parent_name != "ActionPrototype" and current_type == "bytes" ->
        "Wrapper"

      parent_name == "Error" and field_name == "code" ->
        "ErrorCode"

      true -> current_type
    end
  end

  def extra_headers() do
    [{"import \"error.proto\";\n", 0}]
  end

end

GenerateProtoFile.run(System.argv())
